File: Java Implementation/Graph.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

public class Graph implements GraphInterface{

	private HashMap<Integer, Node> nodes;
	
	public Graph() {
		this.nodes = new HashMap<Integer, Node>();
	}

	public Node addNode(Node node) {
		int nodeId = node.getId();
		if(!(this.nodes.containsKey(nodeId))) {
			Node newNode = new Node(nodeId);
			this.nodes.put(nodeId, newNode);
		}
		return this.nodes.get(nodeId);
	}
	
	public Node addNode(int nodeId) {
		if(!(this.nodes.containsKey(nodeId))) {
			Node newNode = new Node(nodeId);
			this.nodes.put(nodeId, newNode);
		}
		return this.nodes.get(nodeId);		
	}
	
	public Node getNode(int nodeId) {
		if(this.nodes.containsKey(nodeId)) {
			return this.nodes.get(nodeId);
		}
		return null;
	}
	
	public ArrayList<Node> nodes(){
		ArrayList<Node> nodes = new ArrayList<>();
		
		for(int nodeId : this.nodes.keySet()) {
			nodes.add(this.nodes.get(nodeId));
		}
		return nodes;
	}

	public ArrayList<Node> sortedNodes(){
		ArrayList<Integer> sortedIDs = new ArrayList<>();
		ArrayList<Node> sortedNodes = new ArrayList<>();
		
		for(int nodeId : this.nodes.keySet()) {
			sortedIDs.add(nodeId);
		}
		sortedIDs.sort(null);
		for(int nodeId : sortedIDs) {
			sortedNodes.add(this.nodes.get(nodeId));
		}
		return sortedNodes;
	}

	
	public int getNumberOfNodes() {
		return this.nodes.size();
	}
	
	
	public boolean hasEdge(Node node1, Node node2) {
		return node1.hasEdge(node2);
	}
	
	public boolean hasEdge(int nodeId1, int nodeId2) {
		return this.hasEdge(getNode(nodeId1), getNode(nodeId2));
	}

	public void addEdge(Node node1, Node node2) {
//		to make sure node1 and node2 exist in the graph.
		Node n1 = this.addNode(node1);
		Node n2 = this.addNode(node2);
		
//		if the new is a self-loop.
		if(n1.equals(n2)) {
			n1.addConnection(n2);
			return;
		}
		n1.addConnection(n2);
		n2.addConnection(n1);
	}
	
	public void addEdge(int nodeId1, int nodeId2) {
//		to make sure node1 and node2 exist in the graph.
		Node n1 = this.addNode(nodeId1);
		Node n2 = this.addNode(nodeId2);
		
//		if the new is a self-loop.
		if(n1.equals(n2)) {
			n1.addConnection(n2);
			return;
		}
		n1.addConnection(n2);
		n2.addConnection(n1);		
	}

	public void addEdge(Node node1, Node node2, double weight) {
//		to make sure node1 and node2 exist in the graph.
		Node n1 = this.addNode(node1);
		Node n2 = this.addNode(node2);
		
//		if the new is a self-loop.
		if(n1.equals(n2)) {
			n1.addConnection(n2, weight);
			return;
		}
		n1.addConnection(n2, weight);
		n2.addConnection(n1, weight);
	}

	public void addEdge(int nodeId1, int nodeId2, double weight) {
//		to make sure node1 and node2 exist in the graph.
		Node n1 = this.addNode(nodeId1);
		Node n2 = this.addNode(nodeId2);
		
//		if the new is a self-loop.
		if(n1.equals(n2)) {
			n1.addConnection(n2, weight);
			return;
		}
		n1.addConnection(n2, weight);
		n2.addConnection(n1, weight);
	}

	public ArrayList<Edge> edges(){
		ArrayList<Edge> allEdges= new ArrayList<>();
		HashMap<Integer, Double> neighbors;
		
		for(Node node : this.nodes()) {
			neighbors = node.neighbors();
			for(int neighborId : neighbors.keySet()) {
				if(node.getId() <= neighborId) {
					allEdges.add(new Edge(node, this.getNode(neighborId), neighbors.get(neighborId)));
				}
			}
//			if(neighbors.keySet().contains(node.getId())) {
//				allEdges.add(new Edge(node, node, neighbors.get(node.getId())));
//			}
		}
		return allEdges;
	}
	
	public int getNumberOfEdges() {
		int numberOfEdges = 0;
		HashMap<Integer, Double> neighbors;
		for(int nodeId : this.nodes.keySet()) {
			neighbors = this.nodes.get(nodeId).neighbors();
//			if node has a self-loop
			if(neighbors.containsKey(nodeId)) {
				numberOfEdges += 1;
			}
			numberOfEdges += neighbors.size();
		}
		return numberOfEdges / 2;
	}
	
	public Graph copy() {
		Graph duplicate = new Graph();
		HashMap<Integer, Double> neighbors;

		for(int nodeId : this.nodes.keySet()) {
			duplicate.addNode(nodeId);
			neighbors = this.nodes.get(nodeId).neighbors();
			for(int neighborId : neighbors.keySet()) {
				duplicate.addEdge(nodeId, neighborId, neighbors.get(neighborId));
			}
		}		
		return duplicate;
	}
	
	public double size() {
		double totalWeight = 0.0;
		HashMap<Integer, Double> neighbors;

		for(Map.Entry<Integer, Node> node : nodes.entrySet()) {
			neighbors = node.getValue().neighbors();
			if(neighbors.containsKey(node.getKey())) {
				totalWeight += node.getValue().edgeWeight(node.getKey());
			}
			for(Map.Entry<Integer, Double> neighbor : neighbors.entrySet()) {
				totalWeight += neighbor.getValue();
			}
		}
		return totalWeight / 2.0;
	}
	
	public double degree(Node node) {
		return this.degree(node.getId());
	}

	public double degree(int nodeId) {
		double degree = 0.0;
		HashMap<Integer, Double> neighbors = this.nodes.get(nodeId).neighbors();

		if(neighbors.containsKey(nodeId)) {
			degree += neighbors.get(nodeId);
		}
		for(int neighbor : neighbors.keySet()) {
			degree += neighbors.get(neighbor);
		}
		return degree;
	}

	public double edgeWeight(int nodeId1, int nodeId2) {
		Node node1 = this.getNode(nodeId1);
		Node node2 = this.getNode(nodeId2);
		
		if(node1 != null && node2 != null) {
			return node1.edgeWeight(node2);
		}
		return 0.0;
	}
	
	public double edgeWeight(Node node1, Node node2) {
		return this.edgeWeight(node1.getId(), node2.getId());
	}
	
	public void removeNode(Node node) {
		for(int neighborId : node.neighbors().keySet()) {
			this.getNode(neighborId).removeNeighbor(node);
			this.nodes.remove(node.getId());
		}
	}
	
	public void removeNode(int nodeId) {
		this.removeNode(this.getNode(nodeId));
	}
	
	public void removeEdge(Node node1, Node node2) {
		node1.removeNeighbor(node2);
		// if(node1.neighbors().size() == 0) {
		// 	this.nodes.remove(node1.getId());
		// }
		node2.removeNeighbor(node1);
		// if(node2.neighbors().size() == 0) {
		// 	this.nodes.remove(node2.getId());
		// }
	}
	
	public HashSet<Integer> getCommonNeighbors(Node node1, Node node2) {
		HashSet<Integer> commonNeighbors = new HashSet<>();
		HashSet<Integer> neighbors1 = new HashSet<>();
		
		for(int neighborId : node1.neighbors().keySet()) {
			neighbors1.add(neighborId);
		}
		for(int neighborId : node2.neighbors().keySet()) {
			if(neighbors1.contains(neighborId)) {
				commonNeighbors.add(neighborId);
			}
		}
		return commonNeighbors;
	}
}

File: C++ Implementation/Graph.cpp
#include "Graph.hpp"
#include <iostream>
#include <string>
#include <algorithm>
#include <unordered_set>


Graph::Graph() {
    this->nodes = new std::unordered_map<int, Node*>;
}

Graph::Graph(const Graph &_graph) : nodes(_graph.nodes) {}

Graph::~Graph() {
    for(auto node : *nodes) {
        delete node.second;
    }
    delete nodes;
}

Node* Graph::add_node(Node* node) {
    int node_id = node->get_id();
    return add_node(node_id);
}

Node* Graph::add_node(int node_id) {
    auto new_node_it = this->nodes->find(node_id);
    if(new_node_it == this->nodes->end()){
        this->nodes->insert(std::make_pair(node_id, new Node(node_id)));
        return this->nodes->at(node_id);
    }
    return new_node_it->second;
}

Node* Graph::get_node(int node_id) const{
    return this->nodes->at(node_id);
}

std::unordered_map<int, Node*>* Graph::get_nodes() const{
    return this->nodes;
}

int Graph::get_number_of_nodes() const{
    return this->nodes->size();
}

int Graph::get_number_of_edges() const{
    int number_of_edges = 0;
    std::unordered_map<Node*, float>* neighbors;

    for(auto node : *(this->nodes)){
        neighbors = node.second->neighbors();
        if(neighbors->find(node.second) != neighbors->end()){
            number_of_edges++;
        }
        number_of_edges += neighbors->size();
    }
    return number_of_edges / 2;
}

bool Graph::has_edge(Node* node_1, Node* node_2) {
    return node_1->has_edge(node_2);
}

bool Graph::has_edge(int node_id_1, int node_id_2) {
    return this->has_edge(nodes->at(node_id_1), nodes->at(node_id_2));
}

void Graph::add_edge(Node* node_1, Node* node_2) {
    Node* n1 = add_node(node_1);
    Node* n2 = add_node(node_2);

    if(n1->get_id() == n2->get_id()){
        n1->add_connection(n2);
        return;
    }

    n1->add_connection(n2);
    n2->add_connection(n1);
}

void Graph::add_edge(int node_id_1, int node_id_2) {
    Node* n1 = add_node(node_id_1);
    Node* n2 = add_node(node_id_2);

    if(n1->get_id() == n2->get_id()){
        n1->add_connection(n2);
        return;
    }

    n1->add_connection(n2);
    n2->add_connection(n1);
}

void Graph::add_edge(Node* node_1, Node* node_2, float weight) {
    Node* n1 = add_node(node_1);
    Node* n2 = add_node(node_2);

    if(n1->get_id() == n2->get_id()){
        n1->add_connection(n2, weight);
        return;
    }

    n1->add_connection(n2, weight);
    n2->add_connection(n1, weight);
}

void Graph::add_edge(int node_id_1, int node_id_2, float weight) {
    Node* n1 = add_node(node_id_1);
    Node* n2 = add_node(node_id_2);

    if(n1->get_id() == n2->get_id()){
        n1->add_connection(n2, weight);
        return;
    }

    n1->add_connection(n2, weight);
    n2->add_connection(n1, weight);
}

float Graph::size() const{
    float total_weight = 0.0;
    std::unordered_map<Node*, float>* neighbors;

    for(auto node : *(this->nodes)){
        neighbors = node.second->neighbors();
        if(neighbors->find(node.second) != neighbors->end()){
            total_weight += node.second->get_edge_weight(node.second);
        }
        for(auto neighbor : *neighbors){
            total_weight += neighbor.second;
        }
    }
    return total_weight / 2.0;
}

float Graph::degree(Node* node) const{
    return node->degree();
}

float Graph::degree(int node_id) const{
    if(this->nodes->find(node_id) == this->nodes->end()) {
        return 0.0;
    }
    return this->get_node(node_id)->degree();
}

float Graph::get_edge_weight(Node* node_1, Node* node_2) const{
    if(node_1 != NULL && node_2 != NULL){
        return node_1->get_edge_weight(node_2);
    }
    return 0.0;
}

void Graph::remove_node(Node* node) {
    for(auto neighbor : *(node->neighbors())) {
        neighbor.first->remove_neighbor(node);
    }
    this->nodes->erase(node->get_id());
}

void Graph::remove_node(int node_id) {
    remove_node(get_node(node_id));
}

void Graph::remove_edge(Node* node_1, Node* node_2) {
    node_1->remove_neighbor(node_2);
    // if(node_1->neighbors().size() == 0) {
    //     this->nodes->erase(node_1->get_id());
    // }
    node_2->remove_neighbor(node_1);
    // if(node_2->neighbors().size() == 0) {
    //     this->nodes->erase(node_2->get_id());
    // }    
}

void Graph::remove_edge(int node_id_1, int node_id_2) {
    this->remove_edge(this->get_node(node_id_1), this->get_node(node_id_2));
}

int Graph::get_common_neighbors(Node* node_1, Node* node_2) {
    std::unordered_set<Node*> common_neighbors;
    std::unordered_set<Node*> neighbor_ids_node_1;
    for(auto neighbor : *(node_1->neighbors())) {
        neighbor_ids_node_1.insert(neighbor.first);
    }
    for(auto neighbor : *(node_2->neighbors())) {
        if(neighbor_ids_node_1.find(neighbor.first) != neighbor_ids_node_1.end()) {
            common_neighbors.insert(neighbor.first);
        }
    }
    return common_neighbors.size();
}

void Graph::print_info() const{
    std::cout << "Printing graph info:\n";
    std::cout << "\t-This graph has " << std::to_string(get_number_of_nodes()) << " node(s).\n";
    std::cout << "\t-This graph has " << std::to_string(get_number_of_edges()) << " edge(s).\n" << std::endl;
}